### PR TITLE
feat(aws-cdk): Detect presence of EC2 credentials

### LINKED ISSUE #130: Toolkit: bail out early if credentials are not defined
Seems like currently even if there are no credentials defined in the default chain, the toolkit will still execute `Looking up default account ID from STS` which will time out and only then a message is displayed that credentials are not found.

We should be able to bail out early if there are no credentials defined in the chain (we should make sure that plugins can still do their job)

### PR DESCRIPTION
**Pull request by rix0rrr**
Automatically detect whether we're on an EC2 instance and only add
looking up metadata credentials if that appears to be true. Add
`--instance`, `--no-instance` command-line arguments to override
the guessing if it happens to be wrong.

This will fix long hangs for people that happen to be on machines
where the metadata service address happens to be routable or blackholed,
such as observed in #702.

Fixes #130.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/aws-cdk/lib/api/util/sdk.ts [hunk 1]
```diff
@@ -290,4 +331,62 @@ function httpsProxyFromEnvironment(): string | undefined {
...
     return undefined;
-}
\ No newline at end of file
+}
+
+/**
+ * Return whether it looks like we'll have ECS credentials available
+ */
+function hasEcsCredentials() {
+    return (AWS.ECSCredentials.prototype as any).isConfiguredForEcsCredentials();
```
Comment (RomainMuller) [inline packages/aws-cdk/lib/api/util/sdk.ts]: Sweet, some more esoteric code of unholy nature doing something that the JS SDK should expose more directly \o/

File: packages/aws-cdk/lib/api/util/sdk.ts [hunk 2]
```diff
@@ -290,4 +331,62 @@ function httpsProxyFromEnvironment(): string | undefined {
...
+}
+
+/**
+ * Return whether we're on an EC2 instance
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/aws-cdk/lib/api/util/sdk.ts]: The [EC2 documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html) mentions those!

File: packages/aws-cdk/lib/api/util/sdk.ts [hunk 3]
```diff
@@ -290,4 +331,62 @@ function httpsProxyFromEnvironment(): string | undefined {
...
+ */
+async function hasEc2Credentials() {
```
Comment (RomainMuller) [inline packages/aws-cdk/lib/api/util/sdk.ts]: Surely, this is not something that could be included in all the SDKs, right? 🤨

Comment (rix0rrr) [reply packages/aws-cdk/lib/api/util/sdk.ts]: Probably most other ones have semi-sane behavior in this regard already.

**Pull request merged**